src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/ActionHandler/CopyHandler.php

```diff
@@ -0,0 +1,170 @@
+        }
+
+        throw new NotCreatableException($modelId->getDataProviderName());
+    }
+
+    /**
+     * Copy a model by using.
+     *
+     * @param ModelIdInterface  $modelId   The model id.
+     *
+     * @return ModelInterface
+     */
+    public function copy(ModelIdInterface $modelId)
+    {
+        $this->guardNotEditOnly($modelId);
+        $this->guardIsCreatable($modelId);
+
+        $environment  = $this->getEnvironment();
+        $dataProvider = $environment->getDataProvider();
+        $model        = $dataProvider->fetch($dataProvider->getEmptyConfig()->setId($modelId->getId()));
+
+        // We need to keep the original data here.
+        $copyModel = $environment->getController()->createClonedModel($model);
+
+        // Dispatch pre duplicate event.
+        $copyEvent = new PreDuplicateModelEvent($environment, $model);
+        $environment->getEventDispatcher()->dispatch($copyEvent::NAME, $copyEvent);
+
+        // Save the copy.
+        $provider = $this->getEnvironment()->getDataProvider($copyModel->getProviderName());
+        $provider->save($copyModel);
+
+        // FIXME: Copy each language multi language models.
+
+        // Dispatch post duplicate event.
+        $copyEvent = new PostDuplicateModelEvent($environment, $copyModel, $model);
+        $environment->getEventDispatcher()->dispatch($copyEvent::NAME, $copyEvent);
+
+        return $copyModel;
+    }
+
+    /**
+     * Redirect to edit mask.
+     *
+     * @param EnvironmentInterface $environment   The environment.
+     * @param ModelIdInterface     $copiedModelId The model id.
+     *
+     * @return void
+     */
+    protected function redirect($environment, $copiedModelId)
+    {
+        // Build a clean url to remove the copy related arguments instad of using the AddToUrlEvent.
+        $url = new BackendUrlBuilder();
+        $url
+            ->setPath('contao/main.php')
+            ->setQueryParameter('do', $environment->getInputProvider()->getParameter('do'))
+            ->setQueryParameter('table', $copiedModelId->getDataProviderName())
+            ->setQueryParameter('act', 'edit')
+            ->setQueryParameter('id', $copiedModelId->getSerialized());
+
+        $redirectEvent = new RedirectEvent($url->getUrl());
+        $environment->getEventDispatcher()->dispatch(ContaoEvents::CONTROLLER_REDIRECT, $redirectEvent);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function process()
+    {
+        $event = $this->getEvent();
+        if ($event->getAction()->getName() !== 'copy') {
+            return;
+        }
+
+        $environment = $this->getEnvironment();
+        $modelId     = ModelId::fromSerialized($environment->getInputProvider()->getParameter('source'));
+
+        $this->guardValidEnvironment($modelId);
+        // We want a redirect here if not creatable.
+        $this->guardIsCreatable($modelId, true);
+
+        if ($environment->getDataDefinition()->getBasicDefinition()->isEditOnlyMode()) {
```

Comment:
IMO this could also become a guard as weare using it in almost every action:
```php
if ($this->isEditOnlyResponse($this->getEvent())) {
  return;
}
```
```php
/**
 * Check if edit only mode and generate the response then.
 */
protected function isEditOnlyResponse()
{
    if (!$environment->getDataDefinition()->getBasicDefinition()->isEditOnlyMode()) {
        return false;
    }

    $event = new ActionEvent($environment, new Action('edit'));
    $environment->getEventDispatcher()->dispatch(DcGeneralEvents::ACTION, $event);
    $this->getEvent()->setResponse($event->getResponse());

    return true;
}
```